tools/storage/segment.py

```diff
@@ -0,0 +1,303 @@
+               'producer_epoch', 'base_seq', 'record_count'))
+
+
+def read_varlong(buffer, pos, bits=64):
+    """
+    Decodes zig-zag encoded signed integers up to 64 bits long as per utils/vint.h
+    from `buffer` starting at `pos`.
+    Returns (value, next @pos)
+    """
+    signbit = 1 << (bits - 1)
+    mask = (1 << bits) - 1
+    result = 0
+    shift = 0
+    while 1:
+        b = six.indexbytes(buffer, pos)
+        result |= ((b & 0x7f) << shift)
+        pos += 1
+        if not (b & 0x80):
+            result &= mask
+            result = (result ^ signbit) - signbit
+            result = int(result)
+            result = (result >> 1) ^ -(result & 1)
+            return (result, pos)
+        shift += 7
+        if shift >= 64:
+            raise Exception("Too many bytes when decoding varint.")
+
+
+class BytesEncoder(json.JSONEncoder):
+    """ Encodes bytes as str for JSON dumps """
+    def default(self, b):
+        if isinstance(b, bytes):
+            return str(b, encoding="utf-8")
+        return json.JSONEncoder.default(self, b)
+
+
+def to_json(record):
+    """ Default record ecoder """
+    return json.dumps(record.__dict__, cls=BytesEncoder)
+
+
+class CorruptBatchError(Exception):
+    def __init__(self, batch):
+        self.batch = batch
+
+
+class Record:
+    def __init__(self, size, record_attr, timestamp_delta,
+                       offset_delta, key, headers, value):
+        self.size = size
+        self.record_attr = record_attr
+        self.timestamp_delta = timestamp_delta
+        self.offset_delta = offset_delta
+        self.key = key
+        self.headers = headers
+        self.value = value
+
+    @staticmethod
+    def from_bytes(b):
+        """ Deocdes record from bytes """
+        size, v = read_varlong(b, 0)
+        record_attr = b[v]
+        v += 1
+
+        timestamp_delta, v = read_varlong(b, v)
+        offset_delta, v = read_varlong(b, v)
+        key_length, v = read_varlong(b, v)
+
+        key = None
+        if key_length > 0:
+            key = b[v: v + key_length]
+            v += key_length
+
+        value_length, v = read_varlong(b, v)
+        value = b[v:v + value_length]
+        v += value_length
+
+        header_count, v = read_varlong(b, v)
+        headers = {}
+        for _ in range(0, header_count):
+            hkey = None
+            hvalue = None
+            hkey_length, v = read_varlong(b, v)
+            if hkey_length > 0: 
+                hkey = b[v:v + hkey_length]
+                v += hkey_length
+            hvalue_length, v = read_varlong(b, v)
+            if hvalue_length > 0:
+                hvalue = b[v:v + hvalue_length]
+                v += hvalue_length
+            if hkey is not None: 
+                headers[hkey] = hvalue
+        return Record(size, record_attr, timestamp_delta,
+                      offset_delta, key, headers, value)
+
+    @staticmethod
+    def records_from_bytes(bytes, count):
+        records = []
```

Comment:
This would be better as a generator, rather than consuming memory for all records in a list before returning it -- it looks like the caller just iterates over the returned list, so shouldn't noticed the difference.